### PR TITLE
feat: Optimizes path finding

### PR DESCRIPTION
**Pull request by hofmeister**
- Auto-Reorder resources based on connection destinations and types
- Optimize pathfinding and use breadth first instead of A-star
- Add grouping of connections between the same 2 blocks
- Add "portal connections" for providers with lots of connections

**Before:** 
image

**After:**
image

### CODE REVIEW

**Comment by hofmeister** [comment]
Note: Before merging - make way to choose between multiple connections that are on top of eachother.

**Comment by hofmeister** [comment]
Took @andertun's idea and made the clustered connections "multi-laned" so we still can interact with them individually. It's functional and better - but I think we can come up with something fancier when @hcirbor is back in action :)

File: src/planner/Planner.tsx [hunk 1]
```diff
@@ -287,6 +266,17 @@ export const Planner = (props: Props) => {
...
         >
             <PlannerCanvas onCreateBlock={props.onCreateBlock}>
+                <svg>
+                    <defs>
+                        <g id={'svg-portal'}>
+                            <ellipse cx="25" cy="25" rx="5" ry="5" fill={'inherit'} />
+                        </g>
+                        <g id={'svg-portal-reverse'}>
+                            <ellipse cx="25" cy="25" rx="5" ry="5" fill={'inherit'} />
+                        </g>
```
Comment (andertun) [inline src/planner/Planner.tsx:276]: They seems to be the same?

**Review by andertun — COMMENTED** [review]
Found a small visual bug, when hovering the consumer the portal dot on the provider is still green.

image

File: src/planner/components/PlannerConnections.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    nodeSize: PlannerNodeSize;
+    highlightedConnections: string[];
+    connections: ConnectionExtension[];
+    onConnectionMouseEnter?: (context: ActionContext) => void;
+    onConnectionMouseLeave?: (context: ActionContext) => void;
+}
+
+export const PlannerConnections = (props: Props) => {
+    const connectionKeys: { [p: string]: boolean } = {};
+    const providerSeen: { [p: string]: boolean } = {};
```
[the remark this answers is not in the document]
Comment (hofmeister) [reply src/planner/components/PlannerConnections.tsx:49]: These are states used as part of the rendering process itself - not something that needs to survive each render

File: src/planner/Planner.tsx (hunk 1 above)
Comment (hofmeister) [reply src/planner/Planner.tsx]: Removed - was left over from a more complex "portal"

**Comment by hofmeister** [comment]
> Found a small visual bug, when hovering the consumer the portal dot on the provider is still green.
> 
> image

Fixed

File: src/planner/components/PlannerConnections.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (andertun) [reply src/planner/components/PlannerConnections.tsx:49]: Ah, of course - makes sense!

**Review by andertun — APPROVED** [review]

**Pull request merged**